### PR TITLE
Use Angular reactive forms

### PR DESCRIPTION
**Pull request by Tcharl**
Fix https://github.com/jhipster/generator-jhipster/issues/8482

-   Please make sure the below checklist is followed for Pull Requests.

-   [x] [Travis tests](https://travis-ci.org/jhipster/generator-jhipster/pull_requests) are green
-   [x] Tests are added where necessary
-   [x] Documentation is added/updated where necessary
-   [x] Coding Rules & Commit Guidelines as per our [CONTRIBUTING.md document](https://github.com/jhipster/generator-jhipster/blob/master/CONTRIBUTING.md) are followed

### CODE REVIEW

File: generators/client/templates/angular/src/main/webapp/app/account/password-reset/finish/password-reset-finish.component.html.ejs [hunk 1]
```diff
@@ -43,50 +43,50 @@
...
+                <form *ngIf="!success" name="form" role="form" (ngSubmit)="finishReset()" [formGroup]="passwordForm">
                     <div class="form-group">
                         <label class="form-control-label" for="password" jhiTranslate="global.form.newpassword">New password</label>
-                        <input type="password" class="form-control" id="password" name="password" #passwordInput="ngModel"
+                        <input type="password" class="form-control" id="password" name="password"
                                placeholder="{{'global.form.newpassword.placeholder' | translate}}"
-                               [(ngModel)]="resetAccount.password" minlength=4 maxlength=50 required>
-                        <div *ngIf="passwordInput.dirty && passwordInput.invalid">
+                               formControlName="newPassword">
+                        <div *ngIf="passwordForm.get(['newPassword']).invalid && (passwordForm.get(['newPassword']).dirty || passwordForm.get(['newPassword']).touched)">
```
Comment (wmarques) [inline generators/client/templates/angular/src/main/webapp/app/account/password-reset/finish/password-reset-finish.component.html.ejs]: I generally use a helper function like `isFieldInvalid(field: string)` to reduce the complexity within the HTML File.
WDYT about doing this for the error checks?

File: generators/client/templates/angular/src/main/webapp/app/account/password-reset/finish/password-reset-finish.component.ts.ejs [hunk 2]
```diff
@@ -28,28 +30,32 @@ import { PasswordResetFinishService } from './password-reset-finish.service';
...
     error: string;
     keyMissing: boolean;
-    resetAccount: any;
     success: string;
     modalRef: NgbModalRef;
     key: string;
 
+    passwordForm = this.fb.group({
+        newPassword: ['', [Validators.required, Validators.minLength(4), Validators.maxLength(50)]],
+        confirmPassword: ['', [Validators.required, Validators.minLength(4), Validators.maxLength(50)]],
```
Comment (wmarques) [inline generators/client/templates/angular/src/main/webapp/app/account/password-reset/finish/password-reset-finish.component.ts.ejs]: Does this works? I always used `Validators.compose` to use multiple validators...

File: generators/client/templates/angular/src/main/webapp/app/account/register/register.component.ts.ejs [hunk 3]
```diff
@@ -31,51 +32,70 @@ import { Register } from './register.service';
...
     register() {
-        if (this.registerAccount.password !== this.confirmPassword) {
+        let registerAccount = {};
+        const login = this.registerForm.get(['login']).value;
+        const email = this.registerForm.get(['email']).value;
+        const password = this.registerForm.get(['password']).value;
+        if (password !== this.registerForm.get(['confirmPassword']).value) {
             this.doNotMatch = 'ERROR';
         } else {
+            registerAccount = Object.assign(registerAccount, {
```
Comment (wmarques) [inline generators/client/templates/angular/src/main/webapp/app/account/register/register.component.ts.ejs]: Maybe we can use spread operator here:
`registerAccount = {...registerAccount, login, email, password} `

File: generators/client/templates/angular/src/main/webapp/app/account/register/register.component.ts.ejs [hunk 4]
```diff
@@ -31,51 +32,70 @@ import { Register } from './register.service';
...
+            });
             this.doNotMatch = null;
             this.error = null;
             this.errorUserExists = null;
             this.errorEmailExists = null;
 <%_ if (enableTranslation) { _%>
             this.languageService.getCurrent().then((key) => {
-                this.registerAccount.langKey = key;
-                this.registerService.save(this.registerAccount).subscribe(() => {
+                registerAccount = Object.assign(registerAccount, {
```
Comment (wmarques) [inline generators/client/templates/angular/src/main/webapp/app/account/register/register.component.ts.ejs]: same, spread operator

File: generators/client/templates/angular/src/main/webapp/app/account/register/register.component.ts.ejs [hunk 5]
```diff
@@ -31,51 +32,70 @@ import { Register } from './register.service';
...
+                    langKey: key
+                });
+                this.registerService.save(registerAccount).subscribe(() => {
                     this.success = true;
                 }, (response) => this.processError(response));
             });
 <%_ } else { _%>
-            this.registerAccount.langKey = 'en';
-            this.registerService.save(this.registerAccount).subscribe(() => {
+            registerAccount = Object.assign(registerAccount, {
```
Comment (wmarques) [inline generators/client/templates/angular/src/main/webapp/app/account/register/register.component.ts.ejs]: spread operator

File: generators/client/templates/angular/src/main/webapp/app/admin/user-management/user-management-update.component.ts.ejs [hunk 6]
```diff
@@ -32,19 +33,31 @@ export class UserMgmtUpdateComponent implements OnInit {
...
-        private router: Router
+        private router: Router,
+        private fb: FormBuilder
     ) {}
 
     ngOnInit() {
         this.isSaving = false;
         this.route.data.subscribe(({user}) => {
             this.user = user.body ? user.body : user;
+            this.valorizeForm(this.user);
```
Comment (wmarques) [inline generators/client/templates/angular/src/main/webapp/app/admin/user-management/user-management-update.component.ts.ejs]: I would prefer this method named `updateForm`, more clear for me, WDYT ?

File: generators/client/templates/angular/src/main/webapp/app/admin/user-management/user-management-update.component.ts.ejs [hunk 7]
```diff
@@ -57,12 +70,25 @@ export class UserMgmtUpdateComponent implements OnInit {
         <%_ } _%>
     }
 
+    private valorizeForm(user): void {
```
Comment (wmarques) [inline generators/client/templates/angular/src/main/webapp/app/admin/user-management/user-management-update.component.ts.ejs]: Can we have some typing for the user  parameter ?

File: generators/entity-client/templates/angular/src/main/webapp/app/entities/entity-management-update.component.ts.ejs [hunk 8]
```diff
@@ -70,62 +71,133 @@ export class <%= entityAngularName %>UpdateComponent implements OnInit {
...
+<%_ } _%>
+        protected activatedRoute: ActivatedRoute,
+        private fb: FormBuilder
     ) {
     }
 
     ngOnInit() {
         this.isSaving = false;
         this.activatedRoute.data.subscribe(({<%= entityInstance %>}) => {
+            this.valorizeForm(<%= entityInstance %>);
```
Comment (wmarques) [inline generators/entity-client/templates/angular/src/main/webapp/app/entities/entity-management-update.component.ts.ejs]: rename to  `updateForm` if that's ok for you

File: generators/entity-client/templates/angular/src/main/webapp/app/entities/entity-management-update.component.ts.ejs [hunk 9]
```diff
@@ -134,8 +206,25 @@ export class <%= entityAngularName %>UpdateComponent implements OnInit {
         return this.dataUtils.openFile(contentType, field);
     }
 
-    setFileData(event, entity, field, isImage) {
-        this.dataUtils.setFileData(event, entity, field, isImage);
+    setFileData(event, field: string, isImage) {
```
Comment (wmarques) [inline generators/entity-client/templates/angular/src/main/webapp/app/entities/entity-management-update.component.ts.ejs]: Why this change?

File: generators/client/templates/angular/src/main/webapp/app/account/password-reset/finish/password-reset-finish.component.html.ejs (hunk 1 above)
Comment (vishal423) [reply generators/client/templates/angular/src/main/webapp/app/account/password-reset/finish/password-reset-finish.component.html.ejs]: If used at more than one place, I generally create getter methods and access those in template and component class like below:

```
get password(): FormControl {
  return this.passwordForm.get('newPassword');
}
```

// use in template like:
`password.invalid `

File: generators/client/templates/angular/src/main/webapp/app/account/password-reset/finish/password-reset-finish.component.ts.ejs [hunk 10]
```diff
@@ -62,10 +68,12 @@ export class PasswordResetFinishComponent implements OnInit, AfterViewInit {
     finishReset() {
         this.doNotMatch = null;
         this.error = null;
-        if (this.resetAccount.password !== this.confirmPassword) {
+        const password = this.passwordForm.get(['newPassword']).value
+        const confirmPassword = this.passwordForm.get(['confirmPassword']).value
+        if (password !== confirmPassword) {
```
Comment (vishal423) [inline generators/client/templates/angular/src/main/webapp/app/account/password-reset/finish/password-reset-finish.component.ts.ejs]: an alternative could be to write custom validator function

File: generators/client/templates/angular/src/main/webapp/app/admin/user-management/user-management-update.component.ts.ejs [hunk 11]
```diff
@@ -71,6 +97,16 @@ export class UserMgmtUpdateComponent implements OnInit {
         }
     }
 
+    private updateUser(user): void {
+        user.login = this.editForm.get(['login']).value;
```
Comment (vishal423) [inline generators/client/templates/angular/src/main/webapp/app/admin/user-management/user-management-update.component.ts.ejs]: if form and model structure matches, then, you can get model object directly using `this.editForm.value`

File: generators/client/templates/angular/src/main/webapp/app/account/password-reset/finish/password-reset-finish.component.ts.ejs [hunk 12]
```diff
@@ -28,28 +30,32 @@ import { PasswordResetFinishService } from './password-reset-finish.service';
...
-    confirmPassword: string;
     doNotMatch: string;
     error: string;
     keyMissing: boolean;
-    resetAccount: any;
     success: string;
     modalRef: NgbModalRef;
     key: string;
 
+    passwordForm = this.fb.group({
```
Comment (vishal423) [inline generators/client/templates/angular/src/main/webapp/app/account/password-reset/finish/password-reset-finish.component.ts.ejs]: an alternative could be to move form model outside in separate model class and do mapping between dto model and form model using services

File: generators/client/templates/angular/src/main/webapp/app/account/password-reset/finish/password-reset-finish.component.ts.ejs (hunk 2 above)
Comment (Tcharl) [reply generators/client/templates/angular/src/main/webapp/app/account/password-reset/finish/password-reset-finish.component.ts.ejs]: Yes, it works, do you prefer compose?

File: generators/entity-client/templates/angular/src/main/webapp/app/entities/entity-management-update.component.ts.ejs (hunk 9 above)
Comment (Tcharl) [reply generators/entity-client/templates/angular/src/main/webapp/app/entities/entity-management-update.component.ts.ejs]: Because we don't have entity at this stage: I'll first try to make it work in a generated project, then contribute to ng-jhipster with a new method.

File: generators/client/templates/angular/src/main/webapp/app/admin/user-management/user-management-update.component.ts.ejs (hunk 11 above)
Comment (Tcharl) [reply generators/client/templates/angular/src/main/webapp/app/admin/user-management/user-management-update.component.ts.ejs]: do you mean that I could use user = this.editForm.value in this case? if it is, this is AWESOME!

File: generators/client/templates/angular/src/main/webapp/app/account/password-reset/finish/password-reset-finish.component.ts.ejs (hunk 12 above)
Comment (Tcharl) [reply generators/client/templates/angular/src/main/webapp/app/account/password-reset/finish/password-reset-finish.component.ts.ejs]: I wanted to stay coherent for form interaction between client and entity classes.
Generating a new model for form is a much time consuming and don't add much value (at least on my side, where forms are 3 to 5 attributes): may be doing this step by step if it is asked and voted (sorry in advance for those who will migrate)?

File: generators/client/templates/angular/src/main/webapp/app/admin/user-management/user-management-update.component.ts.ejs (hunk 11 above)
Comment (vishal423) [reply generators/client/templates/angular/src/main/webapp/app/admin/user-management/user-management-update.component.ts.ejs]: yes

File: generators/client/templates/angular/src/main/webapp/app/account/password-reset/finish/password-reset-finish.component.ts.ejs (hunk 2 above)
Comment (wmarques) [reply generators/client/templates/angular/src/main/webapp/app/account/password-reset/finish/password-reset-finish.component.ts.ejs]: No this way looks clearer for me so that's ok 😄

File: generators/client/templates/angular/src/main/webapp/app/admin/user-management/user-management-update.component.ts.ejs (hunk 11 above)
Comment (wmarques) [reply generators/client/templates/angular/src/main/webapp/app/admin/user-management/user-management-update.component.ts.ejs]: Yes if the model and the form are the same structure it works well :) good catch @vishal423

File: generators/client/templates/angular/src/main/webapp/app/account/password-reset/finish/password-reset-finish.component.html.ejs (hunk 1 above)
Comment (wmarques) [reply generators/client/templates/angular/src/main/webapp/app/account/password-reset/finish/password-reset-finish.component.html.ejs]: Not a bad idea but I would prefer a wrapper to check invalid, dirty and touched and then return a boolean

Comment (vishal423) [reply generators/client/templates/angular/src/main/webapp/app/account/password-reset/finish/password-reset-finish.component.html.ejs]: wrapper can also leverage getter() functions. Point is to reduce complexity from templates (at moment, I see `passwordForm.get(['newPassword'])` code is repeated at more than 5+ places in this template and any future change in property name or object structure would require modification at all these places.

Comment (wmarques) [reply generators/client/templates/angular/src/main/webapp/app/account/password-reset/finish/password-reset-finish.component.html.ejs]: You're right, but I would do a wrapper which takes in a parameter `fieldName` (like `newPassword`) and returns a boolean.
With getters we couldn't do that way, we would need to do a `isInvalid` method per field, if I understood you correctly

Comment (Tcharl) [reply generators/client/templates/angular/src/main/webapp/app/account/password-reset/finish/password-reset-finish.component.html.ejs]: I'm not fan of wrapper: imagine a MOA want to show that some fields are mandatory (errored) without the end user to enter on the field.
If it's a sine qua non feature I'll do, but I would personally prefer without it

File: generators/client/templates/angular/src/main/webapp/app/account/password-reset/finish/password-reset-finish.component.ts.ejs (hunk 10 above)
Comment (Tcharl) [reply generators/client/templates/angular/src/main/webapp/app/account/password-reset/finish/password-reset-finish.component.ts.ejs]: Yes, but I wouldn't change the original code so much

File: generators/client/templates/angular/src/main/webapp/app/admin/user-management/user-management-update.component.ts.ejs (hunk 11 above)
Comment (Tcharl) [reply generators/client/templates/angular/src/main/webapp/app/admin/user-management/user-management-update.component.ts.ejs]: Tried but it only works if all the fields are identical: unfortunately, there are case where fields are undefined so that it fails...

**Comment by Tcharl** [comment]
Everything's perfectly working on local: does someone knows why travis is failing?
May be because of Linux and date: does someone with a linux machine can test?

Thank you in advance

**Comment by pascalgrimaud** [comment]
I'm closing and reopening to restart the build

**Comment by wmarques** [comment]
@Tcharl looks like you have issues with e2e tests + compilation errors `ERROR in : Can't bind to 'translateValues' since it isn't a known property of 'h2'. ("<div> <div class="row justify-content-center"> <div class="col-md-8"> <h2 [ERROR ->][translateValues]="{username: settingsForm.get(['login']).value}" *ngIf="settingsForm.get(['login']).")`

**Comment by Tcharl** [comment]
Everything's good to me: do you want something more in order to merge?

File: generators/client/templates/angular/src/main/webapp/app/account/password-reset/finish/password-reset-finish.component.html.ejs (hunk 1 above)
Comment (wmarques) [reply generators/client/templates/angular/src/main/webapp/app/account/password-reset/finish/password-reset-finish.component.html.ejs]: Ok, let's keep it that way, it is indeed more flexible

File: generators/client/templates/angular/src/main/webapp/app/account/password-reset/init/password-reset-init.component.ts.ejs [hunk 13]
```diff
@@ -50,7 +51,7 @@ export class PasswordResetInitComponent implements OnInit, AfterViewInit {
         this.error = null;
         this.errorEmailNotExists = null;
 
-        this.passwordResetInitService.save(this.resetAccount.email).subscribe(() => {
+        this.passwordResetInitService.save(this.resetRequestForm.get(['email']).value).subscribe(() => {
```
Comment (wmarques) [inline generators/client/templates/angular/src/main/webapp/app/account/password-reset/init/password-reset-init.component.ts.ejs]: Can't we just use `.get('email')`?

File: generators/client/templates/angular/src/main/webapp/app/account/password/password.component.html.ejs [hunk 14]
```diff
@@ -32,63 +32,62 @@
...
-                    <input type="password" class="form-control" id="currentPassword" name="currentPassword" #currentPasswordInput="ngModel"
+                    <input type="password" class="form-control" id="currentPassword" name="currentPassword"
                            placeholder="{{'global.form.currentpassword.placeholder' | translate}}"
-                           [(ngModel)]="currentPassword" required>
-                    <div *ngIf="currentPasswordInput.dirty && currentPasswordInput.invalid">
+                           formControlName="currentPassword">
+                    <div *ngIf="passwordForm.get(['currentPassword']).invalid && (passwordForm.get(['currentPassword']).dirty || passwordForm.get(['currentPassword']).touched)" >
                         <small class="form-text text-danger"
-                               *ngIf="currentPasswordInput.errors.required" jhiTranslate="global.messages.validate.newpassword.required">
+                               *ngIf="passwordForm.get(['currentPassword']).errors.required" jhiTranslate="global.messages.validate.newpassword.required">
```
Comment (wmarques) [inline generators/client/templates/angular/src/main/webapp/app/account/password/password.component.html.ejs]: use the string parameter instead of array one: `.get('currentPassword')`

**Comment by wmarques** [comment]
@Tcharl looks good, only a comment about the usage of the `.get()` function, I think you can directly use a string parameter instead of an array one

File: generators/client/templates/angular/src/main/webapp/app/account/password/password.component.html.ejs (hunk 14 above)
Comment (Tcharl) [reply generators/client/templates/angular/src/main/webapp/app/account/password/password.component.html.ejs]: Done for each fields!

**Comment by Tcharl** [comment]
@wmarques Done, there was many array to remove and it's now better from a end user point of view.
React tests are failing on ci, I don't think that it's my PR fault.

**Comment by wmarques** [comment]
@Tcharl LGTM, thanks alot !

File: generators/client/templates/angular/src/main/webapp/app/admin/user-management/user-management-update.component.ts.ejs (hunk 11 above)
Comment (yelhouti) [reply generators/client/templates/angular/src/main/webapp/app/admin/user-management/user-management-update.component.ts.ejs]: @Tcharl I know this is probably old, but could you please specify which field where problematic, since I already implemented all of this in a blueprint. but maybe I missed something.
